path_login.go

```diff
@@ -396,12 +398,48 @@ func (b *azureAuthBackend) verifyResource(ctx context.Context, subscriptionID, r
 		for _, userIdentity := range resp.Identity.UserAssignedIdentities {
 			principalIDs[convertPtrToString(userIdentity.PrincipalID)] = struct{}{}
 		}
+	default:
+		// in some cases (particularly WIF), a vm/vmss/resource_id might not be provided, in that case
+		// we'll try to authenticate by matching the claim's app_id to the list of managed identities
+		// (see the comment below on that)
+		if claims.AppID == "" {
+			return errors.New("one of vm_name, vmss_name, resource_id, or an app_id JWT claim must be provided")
+		}
 	}
 
 	// Ensure the token OID is the principal id of the system-assigned identity
 	// or one of the user-assigned identities
 	if _, ok := principalIDs[claims.ObjectID]; !ok {
-		return errors.New("token object id does not match expected identities")
+		// if it isn't, check the appID and see if _that_ exists. In some cases, particularly WIF (workload identity
+		// federation), there is no principal that matches the incoming ObjectID. In this case, we can still validate
+		// by checking the appID against the list of managed identities. (The appID is valid for use with authorizing
+		// claims, per https://learn.microsoft.com/en-us/azure/active-directory/develop/access-tokens#payload-claims)
+		if claims.AppID != "" {
+			clientIDs := map[string]struct{}{}
+			c, err := b.provider.MSIClient(subscriptionID) // this is the second time we're calling this, is there a way to reuse that?
+			if err != nil {
+				return fmt.Errorf("failed to create client to retrieve app ids: %w", err)
+			}
+			pager := c.NewListByResourceGroupPager(resourceGroupName, &armmsi.UserAssignedIdentitiesClientListByResourceGroupOptions{})
+			for pager.More() {
+				page, err := pager.NextPage(ctx)
+				if err != nil {
+					return fmt.Errorf("failed to advance page: %w", err)
+				}
+				for _, identity := range page.Value {
+					if identity.Properties != nil && identity.Properties.ClientID != nil {
+						clientIDs[*identity.Properties.ClientID] = struct{}{}
+					}
+				}
+			}
+
+			if _, ok := clientIDs[claims.AppID]; !ok {
+				return errors.New("neither token object id nor token app id match expected identities")
+			}
+		} else {
```

Comment:
Nit: You could probably switch the statement so we we check for `if claims.AppID = "" { return error }` first so the happy path doesn't get this extra nesting.

Reply:
good call!